consensus/core/src/core.rs

```diff
@@ -852,6 +952,79 @@ impl Core {
         true
     }
 
+    // Try to decide which of the certified commits will have to be committed next respecting the `limit`. If provided `limit` is zero, it will panic.
+    // The function returns the list of decided leaders and updates in place the remaining certified commits. If empty vector is returned, it means that
+    // there are no certified commits to be committed as `certified_commits` is either empty or all of the certified commits are already committed.
+    #[tracing::instrument(skip_all)]
+    fn try_decide_certified(
+        &mut self,
+        certified_commits: &mut Vec<CertifiedCommit>,
+        limit: usize,
+    ) -> Vec<(DecidedLeader, CertifiedCommit)> {
+        // If GC is disabled then should not run any of this logic.
+        if !self.dag_state.read().gc_enabled() {
+            return Vec::new();
+        }
+
+        assert!(limit > 0, "limit should be greater than 0");
+
+        let last_commit_index = self.dag_state.read().last_commit_index();
+
+        // If there are certified committed leaders, check that the first certified committed leader which is higher than the last decided one has no gaps.
+        while !certified_commits.is_empty() {
+            let certified_commit = certified_commits
+                .first()
+                .expect("Synced commits should not be empty");
+            if certified_commit.index() <= last_commit_index {
+                tracing::debug!(
+                    "Skip commit for index {} as it is already committed with last commit index {}",
+                    certified_commit.index(),
+                    last_commit_index
+                );
+                certified_commits.remove(0);
+            } else {
+                // Make sure that the first commit we find is the next one in line and there is no gap.
+                if certified_commit.index() != last_commit_index + 1 {
+                    panic!("Gap found between the certified commits and the last committed index. Expected next commit index to be {}, but found {}", last_commit_index + 1, certified_commit.index());
+                }
+
+                // now break as we want to process the rest of the committed leaders
+                break;
+            }
+        }
+
+        if certified_commits.is_empty() {
+            return Vec::new();
+        }
+
+        let to_commit = if certified_commits.len() >= limit {
```

Comment:
For logic above this line, I think they are better to be around core.rs. For logic below, they are indeed closer to commit rule. But tbh I'm not sure if we should actually move leader schedule commit loop to core as well instead. There are more refactor and we don't have to make things perfect in this PR.

Reply:
Ok I'll keep the logic for now in Core.rs. I agree that overall we could work out some refactoring with the leader schedule loop etc. Let's do that on upcoming PR.